ACE/ace/RAW_Socket.cpp

```diff
@@ -0,0 +1,414 @@
+        }
+    #endif
+  #endif
+
+}
+
+ssize_t
+ACE_RAW_SOCKET::recv (void *buf,
+                      size_t n,
+                      ACE_INET_Addr &addr,
+                      int flags,
+                      const ACE_Time_Value *timeout,
+                      ACE_INET_Addr *to_addr) const
+{
+  ACE_TRACE ("ACE_RAW_SOCKET::recv");
+
+  RECV_EXCEPTION_RETURN();
+
+  if(to_addr == NULL)
+  {
+    return using_common_recv(*this, buf, n, addr, flags);
+  }
+  else
+  {
+    this->get_local_addr (*to_addr);
+    if(!to_addr->is_any())
+    {
+       return using_common_recv(*this, buf, n, addr, flags);
+    }
+
+  }
+
+  struct iovec iov;
+  iov.iov_base = static_cast<char*>(buf);
+  iov.iov_len  = static_cast<int>(n);
+
+  msghdr recv_msg     = {};
+  recv_msg.msg_iov    = &iov;
+  recv_msg.msg_iovlen = 1;
+
+  #ifdef ACE_USE_MSG_CONTROL
+  union control_buffer  cbuf; 
+  if(to_addr != NULL)
+  {
+    fillMsgHdr(recv_msg, addr, &cbuf, sizeof(cbuf));
+  }
+  else
+  {
+    fillMsgHdr(recv_msg, addr, NULL, 0);
+  }
+  #else
+    fillMsgHdr(recv_msg, addr, NULL, 0);
+  #endif
+
+  const ssize_t status = ACE_OS::recvmsg (this->get_handle (),
+                                      &recv_msg,
+                                      flags);
+
+  addr.set_size (recv_msg.msg_namelen);
+  addr.set_type ((static_cast<sockaddr_in *>(addr.get_addr()))->sin_family);
+
+  #ifdef ACE_USE_MSG_CONTROL
+  if(to_addr != NULL)
+    getToAddrFromMsgHdr(recv_msg, *to_addr);
+  #endif
+  
+  return status;
+}
+
+
+ssize_t
+ACE_RAW_SOCKET::send (const void *buf,
+                      size_t n,
+                      const ACE_INET_Addr &addr,
+                      int flags,
+                      const ACE_Time_Value *timeout) const
+{
+  ACE_TRACE ("ACE_RAW_SOCKET::send");
+
+  // Check the status of the current socket.
+  SEND_EXCEPTION_RETURN();
+  
+  sockaddr *saddr = static_cast<sockaddr *>(addr.get_addr ());
+  const int len   = addr.get_size ();
+  return ACE_OS::sendto (this->get_handle (),
+                    (const char *) buf,
+                    n,
+                    flags,
+                    (struct sockaddr *) saddr,
+                    len);
+}
+
+#if defined (ACE_HAS_MSG)
+ssize_t 
+ACE_RAW_SOCKET::send (const iovec iov[],
+                int n,
+                const ACE_INET_Addr &addr,
+                int flags,
+                const ACE_Time_Value *timeout) const
+{
+    ACE_TRACE ("ACE_RAW_SOCKET::send");
+
+    // Check the status of the current socket.
+    SEND_EXCEPTION_RETURN();
+
+    msghdr send_msg     = {};
+
+    send_msg.msg_iov    = const_cast<iovec *>(iov);
+    send_msg.msg_iovlen = n;
+
+    #if defined (ACE_HAS_SOCKADDR_MSG_NAME)
+      send_msg.msg_name = static_cast<struct sockaddr *>(addr.get_addr());
+    #else
+      send_msg.msg_name    = (char *) addr.get_addr ();
+    #endif /* ACE_HAS_SOCKADDR_MSG_NAME */
+      send_msg.msg_namelen = addr.get_size ();
+
+    #if defined (ACE_HAS_4_4BSD_SENDMSG_RECVMSG)
```

Comment:
Use nullptr for pointers

Reply:
Firstly I use nullptr according to the ACE developer guideline.
But when I migrated the RAW Socket into ACE 6.5.x version  and compiled the ACE lib on CentOS7 with gcc4.8, I got some compiling error about nullptr, so I modify them to NULL keyword :(